[EventTiming] keydown PerformanceEntry duration is always too long (waits for keyup event?)

francovs opened the pull request:
#### 1337f64ad12082bf549309e2eaec87e7e99a470a
<pre>
[EventTiming] keydown PerformanceEntry duration is always too long (waits for keyup event?)
<a href="https://bugs.webkit.org/show_bug.cgi?id=300102">https://bugs.webkit.org/show_bug.cgi?id=300102</a>
<a href="https://rdar.apple.com/161911473">rdar://161911473</a>

Reviewed by Ryan Reno.

LocalDOMWindowdispatchPendingEventTimingEntries() now assigns duration
to pending entries with nonzero duration, and doesn&apos;t overwrite
that value when queuing. Entries whose duration would round to zero
are assigned 1ms, avoiding being reassigned; this value is not exposed
to the user, as it is later rounded to zero.

Tests which verify this behavior were added; those introduce a
significant gap between the down and up events to make sure the
deferred down event had its duration set before the dispatch of
the up event:

* LayoutTests/imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html
* LayoutTests/imported/w3c/web-platform-tests/event-timing/gap-pointerdown-pointerup.html

Unfortunately, WKTR is unable to properly send actions chains with
unpaired down/up events in several platforms, which required these
tests to be skipped a lot.

Canonical link: <a href="https://commits.webkit.org/302107@main">https://commits.webkit.org/302107@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d88822e6c90f203301289d1d1e24cdffa93ed53

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127436 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38222 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134512 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78991 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129308 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55607 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97004 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65048 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130384 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38160 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114133 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77484 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36993 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32241 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77886 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108019 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32710 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136997 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41690 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105530 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110482 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105230 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26951 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50707 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29149 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/51674 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54032 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60119 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53266 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55025 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -4510,6 +4510,8 @@ webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/event-retarget.
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/first-input-interactionid-click.html [ Skip ]
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/first-input-interactionid-key.html [ Skip ]
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/first-input-shadow-dom.html [ Skip ]
+webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html [ Skip ]
```
Why are we skipping these tests on iOS?

francovs replied:
Added a comment to explain: in iOS, WKTR does not issue pointerdown/pointerup or keydown/keyup when they are issued in separate action chains, like so:

```JS
await new test_driver.Actions()
  .keyDown('a')
  .send()
// Wait for something
await new test_driver.Actions()
  .keyUp('a')
  .send()
```

I didn't get to the bottom of iOS does this; found is some [ongoing discussion in w3c/webdriver on how this kind of test should behave](https://github.com/w3c/webdriver/issues/1772#issuecomment-3236585600).

rreno reviewed (commented):
Please export the new WPT.

rreno commented on LayoutTests/platform/gtk/TestExpectations:
```diff
@@ -71,6 +71,10 @@ webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/interactionid-a
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/interactionid-aux-pointerdown.html [ Skip ]
 
+# Event timing: timing out for unclear reasons (scheduling?). Events handlers are called but event timing entries aren't produced
+webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html [ Skip ]
```
You should use a different bugzilla to track the work to figure this out.

francovs replied:
Added: https://bugs.webkit.org/show_bug.cgi?id=301110

rreno commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -4545,6 +4545,10 @@ webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/shadow-dom-null
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/toJSON.html [ Skip ]
 
+# Timing out because single down/up events are never issued:
+webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html [ Skip ]
```
ditto, use a different bugzilla

francovs replied:
Opened as: https://bugs.webkit.org/show_bug.cgi?id=301110

rreno commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -928,6 +928,9 @@ imported/w3c/web-platform-tests/webtransport/ [ Skip ]
 webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/interactionid-orphan-pointerup.html [ Skip ]
 
+# Event timing: timing out for unclear reasons (scheduling?). Events handlers are called but event timing entries aren't produced
+webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html [ Skip ]
```
I don't think we should expect this to work on wk1.

francovs replied:
Excluded on https://github.com/WebKit/WebKit/pull/52677

rreno commented on LayoutTests/platform/wpe/TestExpectations:
```diff
@@ -538,6 +538,10 @@ webkit.org/b/244186 imported/w3c/web-platform-tests/event-timing/large-duration-
 webkit.org/b/244186 imported/w3c/web-platform-tests/event-timing/retrievability.html [ Skip ]
 
+# Event timing: timing out for unclear reasons (scheduling?). Events handlers are called but event timing entries aren't produced
+webkit.org/b/214677 imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html [ Skip ]
```
Definitely worth a separate bugzilla for other ports to evaluate when and if they want to enable this feature.

francovs replied:
Created: https://bugs.webkit.org/show_bug.cgi?id=301110

rreno commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2716,15 +2716,24 @@ void LocalDOMWindow::dispatchPendingEventTimingEntries()
     auto renderingTime = performance().nowInReducedResolutionSeconds();
     if (m_pendingPointerDown && !m_pendingPointerDown->duration)
-        m_pendingPointerDown->duration = renderingTime - m_pendingPointerDown->startTime;
+        m_pendingPointerDown->duration = std::max(renderingTime - m_pendingPointerDown->startTime, Seconds(0.001));
```
Use a different `Seconds` constructor. `Seconds::fromMilliseconds` I think?

rreno replied:
I guess that's not a different constructor, technically. But it's a more self-documenting way to construct a duration.

rreno commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2716,15 +2716,24 @@ void LocalDOMWindow::dispatchPendingEventTimingEntries()
+
+    for (auto& keydownEntry : m_pendingKeyDowns) {
+        if (!keydownEntry.value.keyDown.duration)
+            keydownEntry.value.keyDown.duration = std::max(renderingTime - keydownEntry.value.keyDown.startTime, Seconds(0.001));
```
Likewise

rreno commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2716,15 +2716,24 @@ void LocalDOMWindow::dispatchPendingEventTimingEntries()
+            keydownEntry.value.keyDown.duration = std::max(renderingTime - keydownEntry.value.keyDown.startTime, Seconds(0.001));
+
+        if (keydownEntry.value.keyPress && !keydownEntry.value.keyPress->duration)
+            keydownEntry.value.keyPress->duration = std::max(renderingTime - keydownEntry.value.keyPress->startTime, Seconds(0.001));
```
likewise

smfr commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2715,15 +2715,24 @@ void LocalDOMWindow::dispatchPendingEventTimingEntries()
+        m_pendingPointerDown->duration = std::max(renderingTime - m_pendingPointerDown->startTime, Seconds::fromMilliseconds(1));
+
+    for (auto& keydownEntry : m_pendingKeyDowns) {
+        if (!keydownEntry.value.keyDown.duration)
```
Should we be using `std::optional<>` instead of relying on 0 as a magic number for durations?

francovs replied:
Not sure; it wouldn't map as well to the spec text:

> When asked to set event timing entry duration given a [PerformanceEventTiming](https://www.w3.org/TR/event-timing/#performanceeventtiming) timingEntry, a [Window](https://html.spec.whatwg.org/multipage/nav-history-apis.html#window) window, and a [DOMHighResTimeStamp](https://w3c.github.io/hr-time/#idl-def-domhighrestimestamp) renderingTimestamp, perform the following steps:
> 1. If timingEntry’s [duration](https://w3c.github.io/performance-timeline/#dom-performanceentry-duration) attribute value is nonzero, return.

This part of the spec is a bit convoluted and potentially being revisited soon (see https://github.com/w3c/event-timing/issues/159), so I tried to make the mapping to code simple.

francovs committed:
[EventTiming] keydown PerformanceEntry duration is always too long (waits for keyup event?)
https://bugs.webkit.org/show_bug.cgi?id=300102
rdar://161911473

Reviewed by Ryan Reno.

LocalDOMWindowdispatchPendingEventTimingEntries() now assigns duration
to pending entries with nonzero duration, and doesn't overwrite
that value when queuing. Entries whose duration would round to zero
are assigned 1ms, avoiding being reassigned; this value is not exposed
to the user, as it is later rounded to zero.

Tests which verify this behavior were added; those introduce a
significant gap between the down and up events to make sure the
deferred down event had its duration set before the dispatch of
the up event:

* LayoutTests/imported/w3c/web-platform-tests/event-timing/gap-keydown-keyup.html
* LayoutTests/imported/w3c/web-platform-tests/event-timing/gap-pointerdown-pointerup.html

Unfortunately, WKTR is unable to properly send actions chains with
unpaired down/up events in several platforms, which required these
tests to be skipped a lot.

Canonical link: https://commits.webkit.org/302107@main
